File: PartyApp/app/src/main/java/com/example/alex/partyapp/servers/ServersFragment.java
package com.example.alex.partyapp.servers;

import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.alex.partyapp.BaseFragment;
import com.example.alex.partyapp.R;
import com.example.alex.partyapp.databinding.FragmentServersBinding;
import com.example.alex.partyapp.di.App;

import javax.inject.Inject;

public class ServersFragment extends BaseFragment<FragmentServersBinding> {

    private LinearLayoutManager linearLayoutManager;
    private ServersAdapter listingAdapter;

    @Inject
    ServersViewModel.Factory factory;
    private ServersViewModel viewModel;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_servers, container, false);
        return binding.getRoot();
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        App.getAppComponent(this.getContext()).inject(this);

        linearLayoutManager = new LinearLayoutManager(getContext());
        binding.recyclerServers.setLayoutManager(linearLayoutManager);
        listingAdapter = new ServersAdapter();
        binding.recyclerServers.setAdapter(listingAdapter);

        viewModel = ViewModelProviders.of(this, factory).get(ServersViewModel.class);
        viewModel.getServers().observe(this, servers -> {
            listingAdapter.setData(servers);
            listingAdapter.notifyDataSetChanged();
        });

        viewModel.getStatus().observe(this, status -> {
            switch (status){
                case LOGGED_OUT:
                    navigateToLogin();
                    break;
            }
        });

        viewModel.getError().observe(this, error -> {
            switch (error) {
                case NETWORK_ERROR:
                    showMessage(R.string.network_error);
                    break;
                case FAILED:
                    showMessage(R.string.server_error);
                    break;
                case OFFLINE_RESULTS:
                    showMessage(R.string.offline_data);
                    break;
                case AUTHORIZATION_FAILED:
                    showMessage(R.string.authorization_failed);
            }
        });

        binding.buttonLogout.setOnClickListener(view -> viewModel.logout());
        binding.swipeRefresh.setOnRefreshListener(this::navigateToLoading);
    }
}

File: app/src/main/java/com/alex/tesonettesttask/ui/fragments/ServersFragment.kt
package com.alex.tesonettesttask.ui.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.alex.tesonettesttask.R
import com.alex.tesonettesttask.TesonetApplication
import com.alex.tesonettesttask.logic.utils.presentFragment
import com.alex.tesonettesttask.ui.adapters.ServersAdapter
import com.alex.tesonettesttask.ui.presenters.ServerPresenter
import kotlinx.android.synthetic.main.fragment_servers.*
import javax.inject.Inject

class ServersFragment : Fragment() {

    @Inject
    lateinit var serverPresenter: ServerPresenter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        TesonetApplication.instance.mainComponent?.inject(this)
        return inflater.inflate(R.layout.fragment_servers, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    private fun init() {
        val servers = serverPresenter.getCachedServers()
        serverList.adapter = ServersAdapter(servers)
        serverList.layoutManager = LinearLayoutManager(context)
        logoutButton.setOnClickListener {
            serverPresenter.logout()
            activity?.presentFragment(LoginFragment())
        }
    }


}
